[registry] Refactor to use a handler registry

bshelton229 opened the pull request:
This primarily starts a simple handler registry pattern to register source handlers.

It also:
- Switches to pytest to match other recent projects
- Creates a `SnagsbySource` base class whose interface should be shared by all snagsby sources.
- Consolidates some aws only logic into the AwsSource class

Of note the exceptions raise (especially KeyError from registry) will be dealt with in subsequent work to normalize error handling.

melinath commented on snagsby/sources.py:
```diff
@@ -56,41 +63,49 @@ def options(self):
 
-    def get_data(self):
-        return sanitize(self.get_raw_data())
+    def get_boto3_session(self, opts=None):
```
💯 based on my other experiences with boto, I'm really glad to see this already moving to use sessions rather than ad-hoc calls, since it'll make things like adding timeouts much easier later on.

bshelton229 replied:
Yeah, centralizing the configuration of the boto3 session has always worked well for us.

melinath commented on snagsby/sources.py:
```diff
@@ -4,15 +4,22 @@
+try:
+    from json.decoder import JSONDecodeError
+except ImportError:
+    JSONDecodeError = ValueError
```
It would be nice to have a comment here explaining what this is a compatibility shim for, so that it's easier to know when we can safely remove it in the future.

melinath commented on snagsby/sources.py:
```diff
@@ -131,21 +137,30 @@ def get_raw_data(self):
+        return registry.get_handler(source_type)(source)
+    except KeyError:
+        logger.debug('Sources must be one of: {}'.format(
+            ", ".join(registry.get_names())))
```
I'd prefer to have this explicitly return None, since the caller is relying on the behavior, but I'm not sure if our style guide requires it?

bshelton229 commented:
@melinath I addressed your comments, please have another look when you have a moment. As far as dealing with missing handlers, that will be refactored very shortly as part of normalizing exception handling.
